iroh-net/src/magicsock/node_map/node_state.rs

```diff
@@ -295,15 +295,39 @@ impl NodeState {
             (None, Some(relay_url)) => ConnectionType::Relay(relay_url),
             (None, None) => ConnectionType::None,
         };
-        if self.conn_type.update(typ).is_ok() {
-            let typ = self.conn_type.get();
+        if let Ok(prev_typ) = self.conn_type.update(typ.clone()) {
+            // The connection type has changed.
             event!(
                 target: "events.net.conn_type.changed",
                 Level::DEBUG,
                 node = %self.node_id.fmt_short(),
                 conn_type = ?typ,
             );
             info!(%typ, "new connection type");
+
+            // Update some metrics
+            if matches!(prev_typ, ConnectionType::Direct(_)) {
+                inc!(MagicsockMetrics, num_direct_conns_removed);
+            }
+            if matches!(typ, ConnectionType::Direct(_)) {
+                inc!(MagicsockMetrics, num_direct_conns_added);
+            }
+            if matches!(typ, ConnectionType::Relay(_) | ConnectionType::Mixed(_, _))
+                && !matches!(
+                    prev_typ,
+                    ConnectionType::Relay(_) | ConnectionType::Mixed(_, _)
+                )
+            {
+                inc!(MagicsockMetrics, num_relay_conns_added);
+            }
+            if matches!(typ, ConnectionType::Direct(_) | ConnectionType::None)
+                && matches!(
+                    prev_typ,
+                    ConnectionType::Relay(_) | ConnectionType::Mixed(_, _)
+                )
+            {
+                inc!(MagicsockMetrics, num_relay_conns_removed);
+            }
```

Comment:
I wouldn't mind if this were refactored into a `match (prev_typ, typ)` expression.

Reply:
Be careful what you wish for...  This version is huge and contains a whole bunch of states that are unreachable.  Admittedly the `if` version has to be read backwards, where you think about what conditions are needed for a certain state.  So I guess the big match is less of a head-twister.

Reply:
I'd didn't necessarily want it to be an exhaustive match 😅 so yeah I guess I *have* to be careful in what I wish for haha!
Basically, having a last `_ => {}` case would've been fine by me :)